crates/sui-rosetta/src/account.rs

```diff
@@ -10,48 +12,125 @@ use futures::StreamExt;
 use sui_sdk::rpc_types::StakeStatus;
 use sui_sdk::{SuiClient, SUI_COIN_TYPE};
 use sui_types::base_types::SuiAddress;
+use tracing::info;
 
 use crate::errors::Error;
 use crate::types::{
     AccountBalanceRequest, AccountBalanceResponse, AccountCoinsRequest, AccountCoinsResponse,
-    Amount, Coin, SubAccount, SubAccountType, SubBalance,
+    Amount, BlockIdentifier, Coin, SubAccount, SubAccountType, SubBalance,
 };
 use crate::{OnlineServerContext, SuiEnv};
+use std::time::Duration;
 
 /// Get an array of all AccountBalances for an AccountIdentifier and the BlockIdentifier
 /// at which the balance lookup was performed.
 /// [Rosetta API Spec](https://www.rosetta-api.org/docs/AccountApi.html#accountbalance)
-pub async fn balance(
+pub async fn balance_new(
     State(ctx): State<OnlineServerContext>,
     Extension(env): Extension<SuiEnv>,
     WithRejection(Json(request), _): WithRejection<Json<AccountBalanceRequest>, Error>,
 ) -> Result<AccountBalanceResponse, Error> {
     env.check_network_identifier(&request.network_identifier)?;
     let address = request.account_identifier.address;
+    let mut retry_attempts = 5;
     if let Some(SubAccount { account_type }) = request.account_identifier.sub_account {
-        let balances = get_sub_account_balances(account_type, &ctx.client, address).await?;
-        Ok(AccountBalanceResponse {
-            block_identifier: ctx.blocks().current_block_identifier().await?,
-            balances,
-        })
+        while retry_attempts > 0 {
+            let balances_first =
+                get_sub_account_balances(account_type.clone(), &ctx.client, address).await?;
+            let checkpoint1 = ctx
+                .client
+                .read_api()
+                .get_latest_checkpoint_sequence_number()
+                .await?;
+            // Get another checkpoint which is greater than current
+            let mut checkpoint2 = ctx
+                .client
+                .read_api()
+                .get_latest_checkpoint_sequence_number()
+                .await?;
+
+            while checkpoint2 <= checkpoint1 {
+                checkpoint2 = ctx
+                    .client
+                    .read_api()
+                    .get_latest_checkpoint_sequence_number()
+                    .await?;
```

Comment:
I'm not quite i understand the logic behind this, would you be able to explain a bit more its purpose?

Reply:
@caogc we should make a quick public doc on what we wrote and share for everyone's understanding

Reply:
Below is the Rosetta account/balance API implementation:

1. Call the [API](https://docs.sui.io/sui-jsonrpc#suix_getBalance) to get the current live balance for account A and store it as account_balance_1.
2. Call the [API](https://docs.sui.io/sui-jsonrpc#sui_getCheckpoint) to get the latest checkpoint number and store it as checkpoint_1.
3. Call the [API](https://docs.sui.io/sui-jsonrpc#sui_getCheckpoint) to get the latest checkpoint number and store it as checkpoint_2. We must ensure the checkpoint_2 > checkpoint_1, since we want to make sure the network moves forward.
4. Call the [API](https://docs.sui.io/sui-jsonrpc#suix_getBalance) to get the current live balance for account A again and store it as account_balance_2.
5. If account_balance_1 equals account_balance_2, the balance is consistent and can be used. If not, repeat steps 1-4 until the two balance queries return the same result. The result is the (account_balance_2, checkpoint_2).

Reply:
<img width="597" alt="Screenshot 2023-05-09 at 10 52 30 AM" src="https://github.com/MystenLabs/sui/assets/80869623/169e9653-05f5-42fb-87b6-2f2cd475f43e">